Added two new rerun types: hms-hms_accretion and cohems_convbdy_dedt

sgossage opened the pull request:
These rerun types address issues certain models have with numerical convergence; these are described below.

- [x] `rerun_type = hms-hms_accretion` **is done** (but see question at the end of the description below).
- [ ] `rerun_type = co-hems_convbdy_dedt` is **WIP**. I'm trying to narrow down the MLT++ changes that are minimal and optimal.

**`rerun_type = hms-hms_accretion`:**

On the HMS-HMS, this tends to occur with an accretor, near L2 overflow conditions. Relatively high mass transfer rates, super adiabatic conditions (w/ He ionization opacity bumps at the surface of both stars), and super-critical rotation of the accretor tend to conspire to cause MESA's hydrodynamics solver to fail to converge on a solution. 

This limits the accretion rate once the timestep falls below one year, and completely turns it off once the timestep falls below 1 day. This is meant to let the accretor star's structure settle and for the timestep to recover.

I also found that `use_dedt_form_of_energy_equation = .true.` helps further, with 100% of the test models converging when used. However, this is not enabled at the moment. Should it be?

**`rerun_type = co-hems_convbdy_dedt`:**

Using  `use_dedt_form_of_energy_equation = .true.` helps models converge in this regime. Perhaps the most numerous of the models failing to converge are helped by this and also less MLT++ (a larger value of `gradT_excess_f2`). Additionally, a number of models that originally experienced segmentation faults are able to converge when `convective_bdy_weight = 0` is used; apparently this algorithm has something like memory alloc. issues in some cases, so disabling it helps. With 

`use_dedt_form_of_energy_equation = .true.`,  `convective_bdy_weight = 0`, `gradT_excess_f2 = 0.4`, convergence is improved for all test models. However, a certain class of failed models remains: those that appear to experience periodic shell flashes.

sgossage commented:
Right now `rerun_type = co-hems_MLTpOFF` only checks for the 'reach cluster timelimit' termination flag. Is that what most of the CO-HeMS models terminate with? Should this rerun_type check for more flags?

mkruckow commented:
About 1., why do you only check for the time step of star 1? Wouldn't it be better to check the minimum for star 1's and star 2's time step is in the window between 1 day and 1 yr?

About 2., I wouldn't include the `'convective_bdy_weight' : 0`, where it isn't needed. As you said we'd have a different rerun for this. Another question, I'd have here: the models, which have issues with MLT++, don't they may stop with `min_timestep_limit`? Hence, is there a good reason to not sure the `TF1_POOL_ERROR`?

In general: While you named the reruns to be meant for a specific grid type, you didn't check for this while exporting the rerun. Hence, either remove the grid specification in the rerun name, if it could be run on any grid or filter for the grid type when exporting reruns.

sgossage commented:
Thanks, Matthias.

1. I've changed this to check for the min timestep between star 1 and 2 now. Originally I was going off the thinking that the code already checks for the minimum in `$MESA/binary/private/binary_timestep.f90`, but may be better to just do it explicitly to not worry about function call order, etc.
2. I've removed `'convective_bdy_weight' : 0` now. From the test models I have, all of the runs fixed by this rerun type terminated originally due to wall time. Some of the `min_timestep` cases are fixed typically by setting `convective_bdy_weight = 0` and some remain unresolved (those that remain unresolved are helped by `use_dedt_form_of_energy_eqn = .true.`, but it's not a solution for 100% of them).
3. I've removed the grid specification from the rerun names as suggested because I think these could be run in principle on any grid that might need them.

sgossage commented:
Thanks again, I've implemented these suggestions.

Could you clarify a bit for your point on code near line 1500? Are you suggesting we would replace that code or retain it and introduce a switch between this PR's functionality and that part of the code?

mkruckow commented:
Sure, I can be a bit more detailed: I had two things in mind:

1. move the new code there, to have the modifications on the same variable closer together (I'm not sure whether you have been aware of this code part, when you did your implementation.)
2. Because the code parts do modify the same parameter we need to think about, what should happen if both cases are fulfilling at the same time. Usually, the later code will simply overwrite what was modified before. The already existing part was implemented to keep the mass transfer solution more stable close to critical rotation, hence can you tell, whether this is used in the cases you looked into and why it doesn't help or why the stars in your case don't get to critical rotation before running into other issues. There is one more point, the old code turns the `mass_transfer_beta` back to 0, in case it is 1 and there is a low mass transfer rate. Hence, are there cases, where the old code would may prevent your code from being effective?

mkruckow commented:
Please don't forget to add the new rerun types to the documentation.

sgossage commented:
> Sure, I can be a bit more detailed: I had two things in mind:
> 
> 1. move the new code there, to have the modifications on the same variable closer together (I'm not sure whether you have been aware of this code part, when you did your implementation.)
> 2. can you tell, whether this is used in the cases you looked into and why it doesn't help or why the stars in your case don't get to critical rotation before running into other issues. There is one more point, the old code turns the `mass_transfer_beta` back to 0, in case it is 1 and there is a low mass transfer rate. Hence, are there cases, where the old code would may prevent your code from being effective?

I see, thanks. 

RE 1: OK, will do. I think maybe using an `x_logical_ctrl` could be good to enable/disable this feature. Maybe there are cases it shouldn't be used

RE 2: This isn't used in the cases I've tested with; for these cases `x_logical_ctrl(4) = .false.`. My impression is that it's false for the rest of the grid too, but is that the case? I agree that the code mentioned could e.g., turn MT back on when the code I'm implementing wants it to be off (it would be turned on if MT rate <= 1e-7 Msol/yr if `mass_transfer_beta` = 1). That might not be a problem since MT is low, but I think making the new code being introduced conditional on an `x_logical_ctrl` as well could be a good compromise for now, since utilizing both pieces of code is untested at the moment and unintended (but not necessarily conflicting).

mkruckow commented on bin/posydon-run-pipeline:
```diff
@@ -619,6 +623,13 @@ def logic_rerun(grid, rerun_type):
+        termination_flags = TF1_POOL_ERROR #['reach cluster timelimit']
+        new_mesa_flag = {'include_composition_in_eps_grav': '.true.',
+                         'change_initial_lnPgas_flag': '.true.',
+                         'new_lnPgas_flag': '.true.'}
```
Usually, we should a change of more than one inlist parameter into a MESA-inlist PR. The reason is, that all the changes put here will end up in the directory names, which become too long for the file system in case of too many parameters.

sgossage replied:
OK thanks. I've now moved these inlist options to inlist1/inlist2 in a separate branch linked above and updated the branch referenced for this rerun in the pipeline script

I've also added an `x_logical_ctrl(7)` switch to control whether timestep_MTlim is enabled or not.

mkruckow commented on docs/_source/components-overview/post_processing/pipeline/pipeline_steps.rst:
```diff
@@ -193,16 +193,18 @@ the type of the rerun specifying the logic and changes.
+    conv_bdy_weight        caution         it disabled the convective_bdy_weight where this caused segmentation faults (this avoids a bug in the old MESA version r11701)
+    timestep_MTlimit       caution         it linearly decreases mass transfer when MESA takes timesteps less than 1 yr and fully turns it off w/ steps less than 1 day. This helps numerical convergence of some low mass accretors.
+    egrav_Pgas_dXdt_terms  caution         it switches the basic variable MESA uses to solve eps_grav from density to Pgas, and tells MESA to use the composition terms when solving. This helps some He stars experiencing numerical issues.
+    ===================    ==============  ===========
```
Just a small remark, you need to change
`    ===================    ==============  ===========`
to
`    =====================  ==============  ===========`
for all the three lines, otherwise some rst readers won't work, e.g. when checking on [github](https://github.com/POSYDON-code/POSYDON/blob/seth_HMS-HMS_CO-HeMS_rerun_fix/docs/_source/components-overview/post_processing/pipeline/pipeline_steps.rst) the final table is not shown. Only the last column is allowed to have shorter table markers.

mkruckow commented on bin/posydon-run-pipeline:
```diff
@@ -482,9 +482,9 @@ def copy_ini_file(grid_type, rerun_metallicity, rerun_type, destination, cluster
+        replace_text = "seth_HMS_HMS_rerun_accretion_fix-f74d4ed021ae67aaa7b89e77da33041550a7d71a"
     elif rerun_type == "egrav_Pgas_dXdt_terms":
-        replace_text = "seth_CO-HeMS_rerun_fix-ecae5eecfc3cecfb20f50b673d0f575bc5a75998"
+        replace_text = "seth_CO_HeMS_rerun_fix-ecae5eecfc3cecfb20f50b673d0f575bc5a75998"
```
It might have been better to rename the branches, to keep our convention of using a `-` between star types in the grid type names.

sgossage replied:
I've renamed the branches to just use the rerun type names for now instead, as there's no real need for me to specify the grids anymore in their names since the implementations are generalized.

As a note though, in `posydon-setup-grid`, the code checks for the branch name and SHA using

```
branch = gitcommit.split('-')[0]
githash = gitcommit.split('-')[1]
```

Should this be changed to e.g.,

```
branch = gitcommit.rsplit('-', 1)[0]
githash = gitcommit.rsplit('-', 1)[-1]
```

This could allow for '-' characters in the branch name so it can be recorded consistently in this code and on github.

mkruckow replied:
I think it would be good to make this change to allow a "-" in the branch name.
I don't see a reason to not just include this change here in this PR-

mkruckow commented:
Just another thought I just had: For the reduction of the accretion efficiency, should we request a minimum mass transfer rate here, e.g. 1e-5 Msun/yr? I think this would already go in the direction to more apply it to systems, which do have significant mass transfer hence although having the accretor to have some material being accreted beforehand and therefore potentially being faster rotating as well? Or is this fix needed for slowing rotating stars already on the very early stage of mass transfer?

mkruckow commented:
Update (March 28): The second rerun type is ready, there are some tries on using a different function for the first one, hopefully completing till next week.

astroJeff reviewed (approved):
Looks good to me

astroJeff commented:
Update (Apr 18): PR consists of links to the git hashes for two new re-runs. The re-run for MT thermal timescale limit is still being perfected. The other re-run (Pgas equations) and not clear how helpful this re-run actually is. It helps at 2 Zsun for HMS+CO RLO grid. Maybe not so much for low Z grids. We will discuss during Monday's collaboration meeting and come up with plan to resolve this.

astroJeff commented:
Update (Apr 25): @sgossage is currently testing these re-runs. Will report back next week

astroJeff commented:
Update (May 9): @sgossage to update the rerun specifics based on presentation at last collaboration meeting.

mkruckow commented on bin/posydon-run-pipeline:
```diff
@@ -481,10 +481,8 @@ def copy_ini_file(grid_type, rerun_metallicity, rerun_type, destination, cluster
         replace_text = "fix_iniHe_pre_rerun5-0e736a853c3c2e522b76299a54aea8f5eea15d08"
-    elif rerun_type == "timestep_MTlimit":
-        replace_text = "seth_rerun_timestep_MTlimit-b94f5dd28d3cacc9fcb00e53bbc3f55f3aecaa00"
-    elif rerun_type == "egrav_Pgas_dXdt_terms":
```
Are we going to dismiss the reruns already done with this on the CO-HeMS gird?

astroJeff reviewed (approved):
Looks good to me

mkruckow reviewed (approved):
I think we can keep it with the more conservative way to call something being really unstable.
